File: src/python/core/infrastructure/embeddings/http_embeddings_client.py
import requests
from typing import List, Dict, Any, Tuple
from ...application.protocols.embeddings_protocol import EmbeddingsClient

ENDPOINT_EMBED = "/embed"
ENDPOINT_INSTRUCTION_PAIRS = "/instruction-pairs"
RESPONSE_VECTORS_KEY = "vectors"
RESPONSE_PAIRS_KEY = "pairs"
RESPONSE_TEXT_KEY = "text"
RESPONSE_METADATA_KEY = "metadata"


class HttpEmbeddingsWrapper:
    def __init__(self, http_client):
        self.http_client = http_client
        
    def embed_documents(self, texts: List[str]) -> List[List[float]]:
        return self.http_client.embed_documents(texts)
    
    def embed_query(self, text: str) -> List[float]:
        return self.http_client.embed_query(text)


class HttpEmbeddingsClient:
    def __init__(self, base_url: str = "http://localhost:8080"):
        self.base_url = base_url.rstrip('/')
        self._embeddings = HttpEmbeddingsWrapper(self)
        
    def embed_documents(self, texts: List[str]) -> List[List[float]]:
        response = requests.post(
            f"{self.base_url}{ENDPOINT_EMBED}",
            json={"texts": texts}
        )
        response.raise_for_status()
        return response.json()[RESPONSE_VECTORS_KEY]
    
    def embed_query(self, text: str) -> List[float]:
        vectors = self.embed_documents([text])
        embedding = vectors[0]
        return embedding
    
    def get_instruction_pairs(self, path: str = None) -> List[Tuple[str, Dict[str, Any]]]:
        url = f"{self.base_url}{ENDPOINT_INSTRUCTION_PAIRS}"
        if path:
            url = f"{url}?path={path}"
            
        response = requests.get(url)
        response.raise_for_status()
        
        pairs_data = response.json()[RESPONSE_PAIRS_KEY]
        return [(pair[RESPONSE_TEXT_KEY], pair[RESPONSE_METADATA_KEY]) for pair in pairs_data]

File: src/dotnet/Core/Infrastructure/Embeddings/HttpEmbeddingsClient.cs
using Rag.Candidates.Core.Application.DTOs.Embeddings;
using Rag.Candidates.Core.Application.Interfaces;
using Rag.Candidates.Core.Domain.Configuration;

namespace Rag.Candidates.Core.Infrastructure.Embeddings;

public sealed class HttpEmbeddingsClient : IEmbeddingsClient
{
    private readonly HttpClient _http;
    private readonly string _base;

    private const string EmbedEndpoint = "/embed";
    private const string InstructionPairsEndpoint = "/instruction-pairs";

    private const string InvalidEmbeddingsResponse = "Invalid embeddings response";
    private const string InvalidInstructionPairsResponse = "Invalid instruction-pairs response";

    public HttpEmbeddingsClient(IHttpClientFactory f, EmbeddingsServiceSettings settings)
    {
        _http = f.CreateClient(nameof(HttpEmbeddingsClient));
        _base = settings.Url;
    }

    public async Task<float[][]> EmbedAsync(IEnumerable<string> texts, CancellationToken ct = default)
    {
        var req = new { texts = texts.ToArray() };
        using var resp = await _http.PostAsJsonAsync($"{_base.TrimEnd('/')}{EmbedEndpoint}", req, ct);
        resp.EnsureSuccessStatusCode();
        var payload = await resp.Content.ReadFromJsonAsync<EmbeddingsResponse>(cancellationToken: ct)
                      ?? throw new InvalidOperationException(InvalidEmbeddingsResponse);
        return payload.Vectors;
    }

    public async Task<(string text, Dictionary<string, object> metadata)[]> GetInstructionPairsAsync(string? path = null, CancellationToken ct = default)
    {
        var url = $"{_base.TrimEnd('/')}{InstructionPairsEndpoint}";
        if (!string.IsNullOrWhiteSpace(path))
        {
            url = $"{url}?path={Uri.EscapeDataString(path)}";
        }
        using var resp = await _http.GetAsync(url, ct);
        resp.EnsureSuccessStatusCode();
        var payload = await resp.Content.ReadFromJsonAsync<InstructionPairsResponse>(cancellationToken: ct)
                      ?? throw new InvalidOperationException(InvalidInstructionPairsResponse);
        return payload.Pairs.Select(p => (p.Text, p.Metadata)).ToArray();
    }
}